torchdata/datapipes/iter/transform/callable.py

```diff
@@ -0,0 +1,129 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Callable, Iterator, List, TypeVar
+
+from torch.utils.data import functional_datapipe, IterDataPipe
+from torch.utils.data.datapipes.utils.common import check_lambda_fn
+
+T_co = TypeVar("T_co", covariant=True)
+
+
+@functional_datapipe("map_batches")
+class BatchMapperIterDataPipe(IterDataPipe[T_co]):
+    r"""
+    Combines elements from the source DataPipe to batches and applies a function
+    over each batch, then flattens the outpus to a single, unnested IterDataPipe
+    (functional name: ``map_batches``).
+
+    Args:
+        datapipe: Source IterDataPipe
+        fn: The function to be applied to each batch of data
+        batch_size: The size of batch to be aggregated from ``datapipe``
+        input_col: Index or indices of data which ``fn`` is applied, such as:
+            - ``None`` as default to apply ``fn`` to the data directly.
+            - Integer(s) is used for list/tuple.
+            - Key(s) is used for dict.
+
+    Example:
+        >>> from torchdata.datapipes.iter import IterableWrapper
+        >>> def fn(batch):
+        >>>     return [d + 1 for d in batch]
+        >>> source_dp = IterableWrapper(list(range(5)))
+        >>> mapped_dp = source_dp.map_batches(fn, batch_size=3)
+        >>> list(mapped_dp)
+        [1, 2, 3, 4, 5]
+
+    Notes:
+        Compared with ``map``, the reason that ``map_batches`` doesn't take
+        ``output_col`` argument is the size of ``fn`` output is not guaranteed
+        to be the same as input batch. With different size, this operation cannot
+        assign data back to original data structure.
+
+        And, this operation is introduced based on the use case from `TorchText`.
+        A pybinded C++ vectorized function can be applied for efficiency.
+    """
+    datapipe: IterDataPipe
+    fn: Callable
+    batch_size: int
+
+    def __init__(
+        self,
+        datapipe: IterDataPipe,
+        fn: Callable,
+        batch_size: int,
+        input_col=None,
+    ) -> None:
+        self.datapipe = datapipe
+
+        check_lambda_fn(fn)
+        self.fn = fn  # type: ignore[assignment]
+
+        assert batch_size > 0, "Batch size is required to be larger than 0!"
+        self.batch_size = batch_size
+        self.input_col = input_col
+
+    def _apply_fn(self, batch):
+        if self.input_col is None:
+            return self.fn(batch)
+
+        if isinstance(self.input_col, (list, tuple)):
+            args = [[data[idx] for idx in self.input_col] for data in batch]
+        else:
+            args = [data[self.input_col] for data in batch]
+        return self.fn(args)
+
+    def __iter__(self) -> Iterator[T_co]:
+        batch: List = []
+        for d in self.datapipe:
```

Comment:
nit: can this be done by `.batch(batch_size).map(...)`?

Reply:
hhhh, totally make sense to me. 😃

Reply:
After testing, `.batch(batch_size).map()` is not equal to `.map_batches()` when `input_col` is specified because `map` would treat input as a single data structure rather than a batch. Then, the `input_col` would be applied at batch level rather than data level.
Let's say our input is `[(0,1), (2,3), (3,4), (5,6), (7,8), (9,10)]` and batch size as 3. With `input_col` as 1, the inputs sent into `fn` are different between the two implementations.
- `.batch().map()`: Inputs are `(2, 3)` and `(7, 8)`
- `.map_batches()`: inputs are `(1, 3, 4)` and `(6, 8, 10)`

Reply:
I see. This is because we have `data[idx]` rather than `batch[idx]` within `_apply_fn`.